libbpf-rs/tests/test.rs

```diff
@@ -369,3 +372,163 @@ fn test_object_reuse_pined_map() {
+fn test_object_ringbuf() {
+    bump_rlimit_mlock();
+
+    let mut obj = get_test_object("ringbuf.bpf.o");
+    let prog = obj
+        .prog("handle__sys_enter_getpid")
+        .expect("error finding program")
+        .expect("failed to find program");
+    let _link = prog.attach().expect("failed to attach prog");
+
+    static mut V1: i32 = 0;
+    static mut V2: i32 = 0;
+
+    fn callback1(data: &[u8]) -> i32 {
+        let mut value: i32 = 0;
+        plain::copy_from_bytes(&mut value, data).expect("Wrong size");
+
+        unsafe {
+            V1 = value;
+        }
+
+        0
+    }
+
+    fn callback2(data: &[u8]) -> i32 {
+        let mut value: i32 = 0;
+        plain::copy_from_bytes(&mut value, data).expect("Wrong size");
+
+        unsafe {
+            V2 = value;
+        }
+
+        0
+    }
+
+    // Test trying to build without adding any ringbufs
+    // Can't use expect_err here since RingBuffer does not implement Debug
+    let builder = libbpf_rs::RingBufferBuilder::new();
+    assert!(
+        builder.build().is_err(),
+        "Should not be able to build without adding at least one ringbuf"
+    );
+
+    // Test building with multiple map objects
+    let mut builder = libbpf_rs::RingBufferBuilder::new();
+
+    // Add a first map and callback
+    let map1 = obj
+        .map("ringbuf1")
+        .expect("Error getting ringbuf1 map")
+        .expect("Failed to get ringbuf1 map");
+
+    builder.add(map1, callback1).expect("Failed to add ringbuf");
+
+    // Add a second map and callback
+    let map2 = obj
+        .map("ringbuf2")
+        .expect("Error getting ringbuf2 map")
+        .expect("Failed to get ringbuf2 map");
+
+    builder.add(map2, callback2).expect("Failed to add ringbuf");
+
+    let mgr = builder.build().expect("Failed to build");
+
+    // Call getpid to ensure the BPF program runs
+    unsafe { libc::getpid() };
+
+    // This should result in both callbacks being called
+    mgr.consume().expect("Failed to consume ringbuf");
+
+    // Our values should both reflect that the callbacks have been called
+    unsafe { assert_eq!(V1, 1) };
+    unsafe { assert_eq!(V2, 2) };
+
+    // Reset both values
+    unsafe { V1 = 0 };
+    unsafe { V2 = 0 };
+
+    // Call getpid to ensure the BPF program runs
+    unsafe { libc::getpid() };
+
+    // This should result in both callbacks being called
+    mgr.poll(Duration::from_millis(100))
+        .expect("Failed to poll ringbuf");
+
+    // Our values should both reflect that the callbacks have been called
+    unsafe { assert_eq!(V1, 1) };
+    unsafe { assert_eq!(V2, 2) };
+}
+
+#[test]
+fn test_object_ringbuf_closure() {
+    bump_rlimit_mlock();
+
+    let mut obj = get_test_object("ringbuf.bpf.o");
+    let prog = obj
+        .prog("handle__sys_enter_getpid")
+        .expect("error finding program")
+        .expect("failed to find program");
+    let _link = prog.attach().expect("failed to attach prog");
+
+    let mut v1 = vec![];
```

Comment:
Maybe use a `channel` ( https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html ) and send something through from the callback to the test?

Reply:
Sure, I'll look into this